sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -510,6 +510,14 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         } else {
           e.copy(branches = branches.take(i).map(branch => (branch._1, elseValue)))
         }
+
+      case EqualTo(CaseWhen(branches, elseValue), right)
+          if right.foldable && branches.forall(_._2.foldable) =>
+        (branches.filter(_._2.equals(right)).map(_._1) ++
```

Comment:
`equals` is only well implemented in `Literal`, but the condition we use is `.foldable`. Shall we change the condition to `.isInstanceOf[Literal]` and wait for the constant folding rule before running this rule?